guides/source/tuning_perf_for_deployment.md

```diff
@@ -0,0 +1,158 @@
+
+To set the number of worker processes, you can change the call to the +workers+ method in +config/puma.rb+. Or you can set the +WEB_CONCURRENCY+ environment variable, which will do the same.
+
+### Preloading
+
+Puma creates new workers from a master process. By loading your application code in the master process, you can avoid doing so after creating the worker. This permits sharing more memory across processes.
+
+In a few cases it may not make sense to preload your application. In that case it's possible to turn off application preloading.
+
+Puma preloads your application by default by calling +preload_app!+ in +config/puma.rb+. If you remove this call, your application will not be preloaded.
+
+### Memory Allocators and Configuration
+
+CRuby normally uses your system's default memory allocator. You can switch to another allocator such as [jemalloc](https://github.com/jemalloc/jemalloc) or [tcmalloc](https://github.com/google/tcmalloc). You can also configure your allocator &mdash; e.g. Linux's glibc malloc allows setting MALLOC_ARENA_MAX=1 to significantly reduce memory use.
+
+This guide does not cover nonstandard allocators in significant detail. However, they can be a significant optimization relative to the system's default allocator. Long-running thread-based workers can be prone to memory fragmentation, which will reduce performance after many requests. An allocator like jemalloc can help.
+
+Performance Testing
+-------------------
+
+Settings from "Choosing Default Settings" are much better than using the initial Rails defaults. Your specific application may have unusual needs or benefit from different configuration options. Load testing takes effort, but can give you more benefit than default settings. You should implement reasonable defaults first.
+
+The best way to choose your application's settings is to test the performance of your application with a simulated production workload. You should save your application's load testing code so you can re-run the tests with future versions of your application.
+
+Performance testing is a deep subject. This guide gives only simple guidelines.
+
+### Load Testers
+
+You will need a load testing program to make requests of your application. This can be a dedicated load testing program of some kind, or you can write a small application to make HTTP requests and track how long they take. You should not normally check the time in your Rails logfile. That time is only how long Rails took to process the request. It does not include time taken by the application server.
+
+Sending many simultaneous requests and timing them can be difficult. It is easy to introduce subtle measurement errors. Normally you should use a load testing program, not write your own. Many load testers are simple to use and many excellent load testers are free.
```

Comment:
maybe we can mention k6 or wrk?

Reply:
I intentionally didn't mention any specific load testing programs. I know @rafaelfranca worries about Rails seeming to endorse specific programs when we don't. So that's why this section doesn't mention any load tester by name.

Reply:
I get the point, but AFAIK the guides have other freedom. I've seen many external tools being mentioned:
- modernizr https://guides.rubyonrails.org/form_helpers.html
- many gems here https://guides.rubyonrails.org/debugging_rails_applications.html
- mobility and traco here https://guides.rubyonrails.org/i18n.html

I think it's valuable to point to specific tools in the guides so it's easier for newcomers. But yeah it'd be good to wait for the Core to clarify this point 👍

Reply:
That's fair. And I mentioned application servers freely, basically because I found other guides that did so.